cmake/modules/FindLibCrypto.cmake

```diff
@@ -11,69 +11,78 @@
 #
 #  LibCrypto_FOUND             System has libcrypto, include and library dirs found
 #  LibCrypto_INCLUDE_DIR       The crypto include directories.
-#  LibCrypto_LIBRARY    The crypto library.
+#  LibCrypto_LIBRARY           The crypto library, depending on the value of BUILD_SHARED_LIBS.
+#  LibCrypto_SHARED_LIBRARY    The path to libcrypto.so
+#  LibCrypto_STATIC_LIBRARY    The path to libcrypto.a
+
 find_path(LibCrypto_INCLUDE_DIR
-        NAMES openssl/crypto.h
-        HINTS
-            ${CMAKE_PREFIX_PATH}/include 
-            ${CMAKE_INSTALL_PREFIX}/include
-        )
+    NAMES openssl/crypto.h
+    HINTS
+        ${CMAKE_PREFIX_PATH}/include 
+        ${CMAKE_INSTALL_PREFIX}/include
+    )
+find_library(LibCrypto_SHARED_LIBRARY
+    NAMES libcrypto.so
+    HINTS
+    ${CMAKE_PREFIX_PATH}/build/crypto
+    ${CMAKE_PREFIX_PATH}/build
+    ${CMAKE_PREFIX_PATH}
+    ${CMAKE_PREFIX_PATH}/lib64
+    ${CMAKE_PREFIX_PATH}/lib 
+    ${CMAKE_INSTALL_PREFIX}/build/crypto
+    ${CMAKE_INSTALL_PREFIX}/build
+    ${CMAKE_INSTALL_PREFIX}
+    ${CMAKE_INSTALL_PREFIX}/lib64
+    ${CMAKE_INSTALL_PREFIX}/lib
+    )
+find_library(LibCrypto_STATIC_LIBRARY
+    NAMES libcrypto.a
+    HINTS 
+    ${CMAKE_PREFIX_PATH}/build/crypto
+    ${CMAKE_PREFIX_PATH}/build
+    ${CMAKE_PREFIX_PATH}
+    ${CMAKE_PREFIX_PATH}/lib64
+    ${CMAKE_PREFIX_PATH}/lib   
+    ${CMAKE_INSTALL_PREFIX}/build/crypto
+    ${CMAKE_INSTALL_PREFIX}/build
+    ${CMAKE_INSTALL_PREFIX}
+    ${CMAKE_INSTALL_PREFIX}/lib64
+    ${CMAKE_INSTALL_PREFIX}/lib
+    )
 
 if (BUILD_SHARED_LIBS)
-    find_library(LibCrypto_LIBRARY
-            NAMES libcrypto.so libcrypto.a
-            HINTS
-            ${CMAKE_PREFIX_PATH}/build/crypto
-            ${CMAKE_PREFIX_PATH}/build
-            ${CMAKE_PREFIX_PATH}
-            ${CMAKE_PREFIX_PATH}/lib64
-            ${CMAKE_PREFIX_PATH}/lib 
-            ${CMAKE_INSTALL_PREFIX}/build/crypto
-            ${CMAKE_INSTALL_PREFIX}/build
-            ${CMAKE_INSTALL_PREFIX}
-            ${CMAKE_INSTALL_PREFIX}/lib64
-            ${CMAKE_INSTALL_PREFIX}/lib
-    )
+    set(LibCrypto_LIBRARY ${LibCrypto_SHARED_LIBRARY})
```

Comment:
What if we’re building a shared lib against a .a with -fPIC? I think this would break.

Reply:
This shouldn't functionally change what would happen in that case. The only thing this change will do is ensure that you always get the .a or .so based on BUILD_SHARED_LIBS, and expose both libs to allow for more explicit linking. Do you think we should also modify the top-level CMakeLists.txt to be more explicit?

Reply:
To be more clear, the problem that happens is that whatever the value of LibCrypto_LIBRARY is becomes a transitive link dependency of anything that calls ```target_link_libraries(target LibCrypto::Crypto)```. As a result, it can end up linking the .so accidentally, even when you're linking s2n statically, if ```BUILD_SHARED_LIBS``` is ```ON``` in the project which consumes s2n as a cmake dependency (because the cmake config files are evaluated in context of the consuming build, which means they flip the linkage).

Reply:
I don't think the transitive link portion is a problem for us per say, because at the top level we don't join the c library builds to the crt build at all. 

It's more the following.

I'm a consumer of s2n (not in a CRT). I want to build libs2n.so and link against libcrypto.a (with -fPIC) how does this support that? It looks like libs2n.so would always try to link against libcrypto.so if it exists.

Reply:
After discussing this offline, we haven't made the problem any worse, it's sort of endemic to CMake and the use of BUILD_SHARED_LIBS. We'd need to solve the problem by teaching s2n to have a parameter for specifying the libcrypto linkage.